### PR TITLE
Persist base mana vectors

### PR DESCRIPTION
**Pull request by acha-bill**
It saves all base mana vectors in storage on shutdown and loads from storage on startup.

### CODE REVIEW

File: packages/mana/persitablebase.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		stringify.StructField("BaseMana1", fmt.Sprint(persistableBaseMana.BaseMana1)),
+		stringify.StructField("EffectiveBaseMana1", fmt.Sprint(persistableBaseMana.EffectiveBaseMana1)),
+		stringify.StructField("BaseMana2", fmt.Sprint(persistableBaseMana.BaseMana2)),
+		stringify.StructField("EffectiveBaseMana2", fmt.Sprint(persistableBaseMana.EffectiveBaseMana2)),
+		stringify.StructField("LastUpdated", fmt.Sprint(persistableBaseMana.LastUpdated)),
+		stringify.StructField("NodeID", persistableBaseMana.NodeID.String()),
+	)
+}
+
+var _ objectstorage.StorableObject = &PersistableBaseMana{}
```
Comment (lzpap) [inline packages/mana/persitablebase.go:42]: what does this line do?

**Review by lzpap — COMMENTED** [review]
LGTM, although I still think that using the object storage is a bit of an overkill for this. We don't need caching for example. Can you think of how to write tests mana persistence?

File: packages/mana/persitablebase.go (hunk 1 above)
Comment (capossele) [reply packages/mana/persitablebase.go:42]: It checks, at compile time, that `PersistableBaseMana` satisfy the `objectstorage.StorableObject` interface

**Comment by acha-bill** [comment]
Yes. We dont need caching. But Object storage is the only interface we have to the database in goshimmer. I think.

For testing, we might have integration test that checks this persistence on shutdown and startup.

Shutdown the node and save its base mana vector.
Start it up again. Then use a new API (e.g get mana vector) to get the vector and assert its not empty or is equal to what was available before shutdown.

File: plugins/mana/plugin.go [hunk 2]
```diff
@@ -127,10 +145,32 @@ func configureEvents() {
 
 func run(_ *node.Plugin) {
 	if err := daemon.BackgroundWorker("Mana", func(shutdownSignal <-chan struct{}) {
-		// TODO: Read base mana vectors from storage
+		readStoredManaVectors()
 		<-shutdownSignal
-		// TODO: write base mana vectors to object storage
+		storeManaVectors()
```
[the remark this answers is not in the document]
Comment (acha-bill) [reply plugins/mana/plugin.go:150]: - yeah. We should shutdown stores.
- yeah. We should prune old values

File: plugins/mana/plugin.go [hunk 3]
```diff
@@ -127,10 +144,46 @@ func configureEvents() {
...
+func storeManaVectors() {
+	for vectorType, baseManaVector := range baseManaVectors {
+		persitables := baseManaVector.ToPersistables()
+		for _, p := range persitables {
+			storages[vectorType].Store(p).Release()
+		}
+	}
+}
+
+func pruneStroages() {
```
Comment (lzpap) [inline plugins/mana/plugin.go:179]:
```suggestion
func pruneStorages() {
```

File: plugins/mana/plugin.go [hunk 4]
```diff
@@ -127,10 +144,46 @@ func configureEvents() {
 
 func run(_ *node.Plugin) {
 	if err := daemon.BackgroundWorker("Mana", func(shutdownSignal <-chan struct{}) {
-		// TODO: Read base mana vectors from storage
+		readStoredManaVectors()
+		pruneStroages()
```
Comment (lzpap) [inline plugins/mana/plugin.go:148]:
```suggestion
		pruneStorages()
```

**Review by lzpap — APPROVED** [review]

**Pull request merged**